src/models/ili9342c.rs

```diff
@@ -0,0 +1,109 @@
+use embedded_graphics_core::{pixelcolor::Rgb565, prelude::IntoStorage};
+use embedded_hal::{blocking::delay::DelayUs, digital::v2::OutputPin};
+
+use crate::{instruction::Instruction, Display, DisplayOptions, Error, Orientation};
+
+use super::{write_command, Model};
+
+/// ILI9342C display with Reset pin
+/// in Rgb565 color mode
+/// Backlight pin is not controlled
+pub struct ILI9342CRgb565;
+
+impl Model for ILI9342CRgb565 {
+    type ColorFormat = Rgb565;
+
+    fn new() -> Self {
+        Self
+    }
+
+    fn init<RST, DELAY, DI>(
+        &mut self,
+        di: &mut DI,
+        rst: &mut Option<RST>,
+        delay: &mut DELAY,
+        options: DisplayOptions,
+    ) -> Result<u8, Error<RST::Error>>
+    where
+        RST: OutputPin,
+        DELAY: DelayUs<u32>,
+        DI: WriteOnlyDataCommand,
+    {
+        match rst {
+            Some(ref mut rst) => self.hard_reset(rst, delay)?,
+            None => write_command(di, Instruction::SWRESET, &[])?,
+        }
+
+        delay.delay_us(120_000);
+
+        init_common(di, delay, options).map_err(|_| Error::DisplayError)
+    }
+
+    fn write_pixels<DI, I>(&mut self, di: &mut DI, colors: I) -> Result<(), DisplayError>
+    where
+        DI: WriteOnlyDataCommand,
+        I: IntoIterator<Item = Self::ColorFormat>,
+    {
+        write_command(di, Instruction::RAMWR, &[])?;
+        let mut iter = colors.into_iter().map(|c| c.into_storage());
+
+        let buf = DataFormat::U16BEIter(&mut iter);
+        di.send_data(buf)
+    }
+
+    fn display_size(&self, orientation: Orientation) -> (u16, u16) {
+        match orientation {
+            Orientation::Portrait(_) | Orientation::PortraitInverted(_) => (320, 240),
+            Orientation::Landscape(_) | Orientation::LandscapeInverted(_) => (240, 320),
+        }
+    }
+}
+
+// simplified constructor for Display
+
+impl<DI, RST> Display<DI, RST, ILI9342CRgb565>
+where
+    DI: WriteOnlyDataCommand,
+    RST: OutputPin,
+{
+    ///
+    /// Creates a new [Display] instance with [ILI9342C] as the [Model]
+    /// *WARNING* Rgb565 only works on non-SPI setups with the ILI9342C!
+    ///
+    /// # Arguments
+    ///
+    /// * `di` - a [DisplayInterface](WriteOnlyDataCommand) for talking with the display
+    /// * `rst` - display hard reset [OutputPin]
+    /// * `model` - the display [Model]
+    ///
+    pub fn ili9342c_rgb565(di: DI, rst: RST) -> Self {
+        Self::with_model(di, Some(rst), ILI9342CRgb565::new())
+    }
+}
+
+// common init for all color format models
+fn init_common<DELAY, DI>(
+    di: &mut DI,
+    delay: &mut DELAY,
+    options: DisplayOptions,
+) -> Result<u8, DisplayError>
+where
+    DELAY: DelayUs<u32>,
+    DI: WriteOnlyDataCommand,
+{
+    let madctl = options.madctl() ^ 0b0000_1000; // this model has flipped RGB/BGR bit;
+
+    write_command(di, Instruction::SLPOUT, &[])?; // turn off sleep
+    write_command(di, Instruction::COLMOD, &[0b0101_0101])?; // 16bit 65k colors
```

Comment:
This should be dynamic based on the color type. It's ok for the 565 variant but I wonder if your 666 case didn't work because of this?

I made the same mistake in the ILI9486 driver (I never could've tested rgb565 because they disabled it in hw for the board I have so I just found this bug).

It can probably be made compile time choice tho based on the color type.

Reply:
Good catch, thank you for pointing that out. Changing to 18bit/262k colours at least made the display not work in a _different_ way using `Rgb666` 😁 I'll keep digging.

Reply:
> 

Hmm this would imply some sort of timing issue. What SPI MODE are you using? Does the display have a CS pin?

Reply:
This ended up being a bit of a red herring.

I was originally testing using `esp32s3-hal` (which is `no_std`), but just as a sanity check I also tested this driver using `esp-idf-hal`/`esp-idf-sys` (which are `std`, a separate project) and this works at all optimization levels. So it appears this issue is somewhere in the `esp-hal` SPI driver.

Reply:
Sounds good! It usually is on the SPI level at that point. Thanks again for the model